File: serverplugin_ts_proj/src/userserver/messages.ts
import { serversdk } from "../helpers/serversdk";

export class message_base {
    constructor() {}

    public static get_type_string(): string {
        return "message_base";
    }

    public static get_type_string_crc(): bigint {
        const type_string = message_base.get_type_string();
        const type_string_crc = serversdk.serverplugin.get_crc32(type_string, type_string.length);
        // type_string_crc = essentials.mod_crc32_z(type_string_crc, type_string, type_string.length);
        return BigInt(type_string_crc);
    }

    public get_type_crc(): bigint {
        return message_base.get_type_string_crc();
    }

    public get_type(): string {
        return message_base.get_type_string();
    }
}

export class rq_msg_user_base extends message_base {
    public pid: string;
    public token: string;

    constructor() {
        super();
        this.pid = "";
        this.token = "";
    }
}

export class rq_msg_user_get extends rq_msg_user_base {
    public device: rq_msg_user_get.device_struct;

    constructor() {
        super();
        this.device = new rq_msg_user_get.device_struct();
    }

    public static get_type_string(): string {
        return "rq_msg_user_get";
    }

    public static get_type_string_crc(): bigint {
        const type_string = rq_msg_user_get.get_type_string();
        const type_string_crc = serversdk.serverplugin.get_crc32(type_string, type_string.length);
        // type_string_crc = essentials.mod_crc32_z(type_string_crc, type_string, type_string.length);
        return BigInt(type_string_crc);
    }

    public override get_type_crc(): bigint {
        return rq_msg_user_get.get_type_string_crc();
    }

    public override get_type(): string {
        return rq_msg_user_get.get_type_string();
    }
}

export namespace rq_msg_user_get {
    export class device_struct {
        public sys_name: string;
        public node_name: string;
        public release: string;
        public arch: string;

        constructor() {
            this.sys_name = "";
            this.node_name = "";
            this.release = "";
            this.arch = "";
        }
    }
}

export class rq_msg_ping {
    public msg: string;
    constructor() {
        this.msg = "";
    }
}

export class FX {
    public static Q_FACTOR: number = 12;
    public static FX_ONE: number = 1 << FX.Q_FACTOR;
    public static FX_TWO: number = FX.FX_ONE + FX.FX_ONE;
}

export class msg_room_config extends message_base {
    public min: number;
    public max: number;
    public betx: number;
    public rewardx: number;
    public allow_after_start: boolean;

    constructor() {
        super();
        this.min = 2;
        this.max = 4;
        this.betx = 0; // Note: betx & rewardx are in fixed-point values
        this.rewardx = FX.FX_TWO;
        this.allow_after_start = false;
    }

    public static get_type_string(): string {
        return "msg_room_config";
    }

    public static get_type_string_crc(): bigint {
        const type_string = msg_room_config.get_type_string();
        const type_string_crc = serversdk.serverplugin.get_crc32(type_string, type_string.length);
        // type_string_crc = essentials.mod_crc32_z(type_string_crc, type_string, type_string.length);
        return BigInt(type_string_crc);
    }

    public override get_type_crc(): bigint {
        return msg_room_config.get_type_string_crc();
    }

    public override get_type(): string {
        return msg_room_config.get_type_string();
    }
}

export class msg_room_config_list extends message_base {
    public configs: msg_room_config[];
    constructor() {
        super();
        this.configs = [];
    }
}

export class res_msg_user_base extends message_base {
    public pid: string;
	public room_list : msg_room_config_list | null;
    constructor() {
        super();
        this.pid='';
        this.room_list = null;
    }
}

// export class res_msg_gservers extends message_base {
//     public gservers : Map<string, string[]>;
//     constructor() {
//         super();
//         this.gservers = new Map<string, string[]>();
//     }
// }
// export class res_msg_gservers_serialisable extends message_base {
//     public gservers : any;
//     constructor(gservers : res_msg_gservers) {
//         super();
//         this.gservers = Object.fromEntries(gservers.gservers);
//     }
// }

export class res_msg_user_get extends res_msg_user_base {
    public last_login: string;
    public user_name: string;
    public token: string;
    public gservers: any;
    constructor() {
        super();
        this.last_login = '';
        this.user_name = '';
        this.token = '';
        this.gservers = null;
    }
 };

File: serverplugin_lua_proj/src/userserver/messages.lua
local serversdk = require("src.helpers.serversdk")

local function define_class(base)
    local class = {}
    class.__index = class

    -- Inherit methods from the base class, if provided
    if base then
        setmetatable(class, { __index = base })
    end

    return class
end

-- Base class: message_base
local message_base = define_class()
function message_base.new()
    local self = setmetatable({}, message_base)
    return self
end

function message_base.get_type_string()
    return "message_base"
end

function message_base.get_type_string_crc()
    local type_string = message_base.get_type_string()
    local type_string_crc = serversdk.serverplugin.get_crc32(type_string, #type_string)
    return tonumber(type_string_crc)
end

function message_base:get_type_crc()
    return message_base.get_type_string_crc()
end

function message_base:get_type()
    return message_base.get_type_string()
end

-- Derived class: rq_msg_user_base
local rq_msg_user_base = define_class(message_base)
function rq_msg_user_base.new()
    local self = setmetatable({}, rq_msg_user_base)
    self.pid = ""
    self.token = ""
    return self
end

-- Derived class: rq_msg_user_get
local rq_msg_user_get = define_class(rq_msg_user_base)
function rq_msg_user_get.new()
    local self = setmetatable({}, rq_msg_user_get)
    self.device = rq_msg_user_get.device_struct.new()
    return self
end

function rq_msg_user_get.get_type_string()
    return "rq_msg_user_get"
end

function rq_msg_user_get.get_type_string_crc()
    local type_string = rq_msg_user_get.get_type_string()
    local type_string_crc = serversdk.serverplugin.get_crc32(type_string, #type_string)
    return tonumber(type_string_crc)
end

function rq_msg_user_get:get_type_crc()
    return rq_msg_user_get.get_type_string_crc()
end

function rq_msg_user_get:get_type()
    return rq_msg_user_get.get_type_string()
end

-- Namespace for device_struct
rq_msg_user_get.device_struct = define_class()
function rq_msg_user_get.device_struct.new()
    local self = setmetatable({}, rq_msg_user_get.device_struct)
    self.sys_name = ""
    self.node_name = ""
    self.release = ""
    self.arch = ""
    return self
end


-- Derived class: res_msg_user_base
local res_msg_user_base = define_class(message_base)
function res_msg_user_base.new()
    local self = setmetatable({}, res_msg_user_base)
    self.pid = ""         -- Default value for pid
    self.room_list = nil  -- Default value for room_list
    return self
end

-- Derived class: res_msg_user_get
local res_msg_user_get = define_class(res_msg_user_base)
function res_msg_user_get.new()
    local self = setmetatable({}, res_msg_user_get)
    self.last_login = ""   -- Default value for last_login
    self.user_name = ""    -- Default value for user_name
    self.token = ""        -- Default value for token
    self.gservers = nil    -- Default value for gservers
    return self
end

return {
    rq_msg_user_get = rq_msg_user_get,
    res_msg_user_get = res_msg_user_get
}
